llama_index/postprocessor/rankGPT_rerank.py

```diff
@@ -0,0 +1,223 @@
+        api_version: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        service_context = service_context or ServiceContext.from_defaults()
+
+        api_key, api_base, api_version = resolve_openai_credentials(
+            api_key=api_key,
+            api_base=api_base,
+            api_version=api_version,
+        )
+        super().__init__(
+            service_context=service_context,
+            top_n=top_n,
+            model=model or service_context.llm.metadata.model_name,
+            temperature=temperature,
+            api_key=api_key,
+            api_base=api_base,
+            api_version=api_version,
+            **kwargs,
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "RankGPTRerank"
+
+    def _postprocess_nodes(
+        self,
+        nodes: List[NodeWithScore],
+        query_bundle: Optional[QueryBundle] = None,
+    ) -> List[NodeWithScore]:
+        if query_bundle is None:
+            raise ValueError("Query bundle must be provided.")
+
+        items = {
+            "query": query_bundle.query_str,
+            "hits": [{"content": node.get_content()} for node in nodes],
+        }
+
+        messages = self.create_permutation_instruction(item=items)
+        permutation = self.run_llm(messages=messages)
+        if permutation.message is not None and permutation.message.content is not None:
+            rerank_ranks = self._receive_permutation(
+                items, str(permutation.message.content)
+            )
+
+            initial_results: List[NodeWithScore] = []
+
+            for idx in rerank_ranks:
+                initial_results.append(
+                    NodeWithScore(node=nodes[idx].node, score=nodes[idx].score)
+                )
+            return initial_results[: self.top_n]
+        else:
+            return nodes[: self.top_n]
+
+    def _get_prefix_prompt(self, query: str, num: int) -> List[ChatMessage]:
+        return [
+            ChatMessage(
+                role="system",
+                content="You are RankGPT, an intelligent assistant that can rank passages based on their relevancy to the query.",
+            ),
+            ChatMessage(
+                role="user",
+                content=f"I will provide you with {num} passages, each indicated by number identifier []. \nRank the passages based on their relevance to query: {query}.",
+            ),
+            ChatMessage(role="assistant", content="Okay, please provide the passages."),
+        ]
+
+    def _get_post_prompt(self, query: str, num: int) -> str:
+        return f"Search Query: {query}. \nRank the {num} passages above based on their relevance to the search query. The passages should be listed in descending order using identifiers. The most relevant passages should be listed first. The output format should be [] > [], e.g., [1] > [2]. Only response the ranking results, do not say any word or explain."
+
+    def _num_tokens_from_messages(
```

Comment:
we have a token counting util 

`from llama_index.utilities.token_counting import TokenCounter`